catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/QueryOperations.java

```diff
@@ -324,10 +247,66 @@ public QueryResponse replaceSourceId(QueryResponse queryResponse) {
         }
         newResponse.setHits(queryResponse.getHits());
         newResponse.closeResultQueue();
-        LOGGER.debug("EXITING: replaceSourceId()");
+        LOGGER.trace("EXITING: replaceSourceId()");
         return newResponse;
     }
 
+    boolean canAccessSource(FederatedSource source, QueryRequest request) {
+        Map<String, Set<String>> securityAttributes = source.getSecurityAttributes();
+        if (securityAttributes.isEmpty()) {
+            return true;
+        }
+
+        Object requestSubject = request.getProperties()
+                .get(SecurityConstants.SECURITY_SUBJECT);
+        if (requestSubject instanceof ddf.security.Subject) {
+            Subject subject = (Subject) requestSubject;
+
+            KeyValueCollectionPermission kvCollection = new KeyValueCollectionPermission(
+                    CollectionPermission.READ_ACTION,
+                    securityAttributes);
+            boolean isPermitted = subject.isPermitted(kvCollection);
+            if (isPermitted) {
+                SecurityLogger.audit("Subject is permitted to access source {}", source.getId());
+            } else {
+                SecurityLogger.audit("Subject is not permitted to access source {}",
```

Comment:
It looks like both of these audits are essentially getting thrown inside of a loop for every request. This could be a potential performance bottleneck if there are a lot of sources. I'd rather see one audit written at the very end and only for the negative case.